### PR TITLE
Corrected a problem with memory

### PR DESCRIPTION
**Pull request by jack-theripper**
I have corrected some inconvenient moments and have up to the end corrected a problem with memory.

fix inheritance support.
fix custom reading function (CURLOPT_READFUNCTION).
fix custom methods.
fix "Expect" and "Accept" header.
fix PATCH with send body.
fix support native PHP stream filters.
fix support follow location (CURLOPT_FOLLOWLOCATION).
fix compatible stream with PSR-7 specifications.
fix out of memory when a large body send of response.
fix out of memory with sendAsyncRequest in some cases.

### CODE REVIEW

**Comment by sagikazarmark** [comment]
:+1: 

Wow, this is super awesome, thank you. Needs a rebase though, and @mekras's review.

**Comment by mekras** [comment]
@jack-theripper, can you split these changes into separate PRs? Also I have some questions and notices.

1. [HttpClient::sendRequest()](https://github.com/php-http/httplug/blob/master/src/HttpClient.php#L27) have only one argument — $request, so [this](https://github.com/php-http/curl-client/pull/19/files#diff-ff73e042e738204c6da009e2ed19f783R94) change is useless.
2. I doubt that Client class should allow descendants. @sagikazarmark, what do you think? My be better mark it final?
3. Please follow [PSR-2](http://www.php-fig.org/psr/psr-2/).

**Comment by jack-theripper** [comment]
@mekras, how to split? This complex correction. What do you mean ?

1. I changed the signature because it is necessary, because it is Curl. It differs from the interface - in it there is nothing bad.

Simple example: how to set the timeout to 1 request?

```php
$client = new Client(MessageFactory, StreamFactory, [
	CURLOPT_TIMEOUT => 10
]);

// send request. timeout = 10
$client->sendRequest($request);

// send other request. timeout = 200
$client->sendRequest($request, [
	CURLOPT_TIMEOUT => 200
]);

// send other request. timeout = 10
$client->sendRequest($request);
```

2 .  The client must be flexible. Maybe, but artificial restrictions - it is bad. Should not be artificial restrictions.

it's all the questions?

**Comment by mekras** [comment]
> how to split? This complex correction. What do you mean ?

At least "replace property visibility" is not necessary for other changes, so can be moved to a separate PR, to be discussed. Setting "Accept" header. May be something else. So this is the reason for splitting PR: clearly define the purpose of each code change.

File: src/Client.php [hunk 1]
```diff
@@ -18,39 +19,42 @@
...
  * @author  Blake Williams <github@shabbyrobe.org>
+ * @author  Dmitry Arhitector <dmitry.arhitector@yandex.ru>
  *
  * @api
  * @since   1.0
  */
 class Client implements HttpClient, HttpAsyncClient
 {
-    /**
+
```
Comment (sagikazarmark) [inline src/Client.php]: Why the extra line?

File: src/Client.php [hunk 2]
```diff
@@ -18,39 +19,42 @@
...
+    protected $handle = null;
 
     /**
      * Simultaneous requests runner
      *
      * @var MultiRunner|null
      */
-    private $multiRunner = null;
+    protected $multiRunner = null;
+
```
Comment (sagikazarmark) [inline src/Client.php]: Why the extra line.

File: src/Client.php [hunk 3]
```diff
@@ -68,29 +72,17 @@ class Client implements HttpClient, HttpAsyncClient
      *
      * @since 1.0
      */
-    public function __construct(
-        MessageFactory $messageFactory,
-        StreamFactory $streamFactory,
-        array $options = []
-    ) {
+    public function __construct(MessageFactory $messageFactory, StreamFactory $streamFactory, array $options = [])
```
Comment (sagikazarmark) [inline src/Client.php]: Please revert this, this is because PSR conforming.

File: src/Client.php [hunk 4]
```diff
@@ -99,27 +91,24 @@ public function __destruct()
      *
      * @since 1.0
      */
-    public function sendRequest(RequestInterface $request)
+    public function sendRequest(RequestInterface $request, array $options = [])
```
Comment (sagikazarmark) [inline src/Client.php]: Please don't add this. Use the constructor.

File: src/Client.php [hunk 5]
```diff
@@ -138,66 +128,111 @@ public function sendRequest(RequestInterface $request)
      *
      * @since 1.0
      */
-    public function sendAsyncRequest(RequestInterface $request)
+    public function sendAsyncRequest(RequestInterface $request, array $options = [])
     {
-        if (!$this->multiRunner instanceof MultiRunner) {
+        if ( ! $this->multiRunner instanceof MultiRunner) {
```
Comment (sagikazarmark) [inline src/Client.php]: Please revert this. No spaces (PSR)

File: src/Client.php [hunk 6]
```diff
@@ -138,66 +128,111 @@ public function sendRequest(RequestInterface $request)
...
 
         return $promise;
     }
 
     /**
+     * Get parser
+     *
+     * @return ResponseParser
+     */
+    public function getResponseParser()
```
Comment (sagikazarmark) [inline src/Client.php]: This isn't necessary.

**Comment by sagikazarmark** [comment]
> I changed the signature because it is necessary, because it is Curl. It differs from the interface - in it there is nothing bad.

Options must be injected in the constructor.

> The client must be flexible. Maybe, but artificial restrictions - it is bad. Should not be artificial restrictions.

Not quite sure what you are trying to say, but I agree with @mekras. Please don't change the visibility. And yes, it should probably be final.

**Comment by jack-theripper** [comment]
@sagikazarmark, "PSR conforming" I didn't pay attention to it, I just set in IDE code style "PSR-2" and auto code align.

you use a "php-http/curl-client" in your code ?

I will ask a few questions:

1. how to install Curl option only for 1 request ?
2. if I use a custom StreamFactory, how to get an object StreamFactory from ResponseParser ? (
if the only access to the variable $client)
3. if I need to change some of the methods (for example "createHeaders"), i make "MyClient extends Client" (it will not work because "private") ?

**Comment by sagikazarmark** [comment]
1. Set up multiple clients if you really need separate options
2. Not sure what you mean. The StreamFactory is injected into the ResponseParser. Why would you need access to it?
3. Not sure how this works in this client, but the behaviour should be 100% configurable. So when adding headers in the client, it either doesn't make sense to change the behaviour or you can do so by some configuration. Also, you can always add custom headers to the request itself. Extending clients is not recommended, we suggest composition over inheritance.

**Comment by jack-theripper** [comment]
@sagikazarmark, There are situations when you need to get control on CreateResponse (not to spend cpu time on reading and writing streams). I work with a very very large body. I need "Except" headers and change the logic, but the client does not support. That is a quick fix, I need "extends Client", but it is forbidden to artificially.

NOTE: Sorry, but it was a mistake to use this client in my project, he is such incomplete.

**Comment by sagikazarmark** [comment]
@jack-theripper although I see your problem, and I think it is important to address it, I am pretty sure that it can be addressed in a proper way.

The Client should be CPU and memory efficient out of the box, I don't see why you should extend it to achieve that. I saw some discussion about except headers as well.

If you allow me a personal comment: crticizing others work and being rude doesn't seem to be the correct behaviour, especially if people want to help you solving your issues, even if they see the solution a bit differently. (Not to mention you are trying to enforce bad design)

I think @mekras did quite a good job with this client and he keeps solving the upcoming issues efficiently. If it doesn't fit your use case, you are free to fork it and add any features you want, if you are not satisfied with our support.

**Comment by jack-theripper** [comment]
@sagikazarmark

> The Client should be CPU and memory efficient out of the box, I don't see why you should extend it to achieve that. I saw some discussion about except headers as well.

because the client out of the box works not effectively. If the client can be inherited, then it will **save developers time**.

I didn't criticize, I just corrected a row of serious problems. 

> fix out of memory. (issues #15) not correctly, leads to hang the server. 

I think I should cancel pull request. I have corrected https://github.com/jack-theripper/curl-client/commit/abbf8fa509763c4fc5b9c460b1f9763364689667  you can use my solution.

**Comment by sagikazarmark** [comment]
@mekras can you take over this and see how we can implement it based on @jack-theripper's solution?

**Comment by mekras** [comment]
@sagikazarmark, I'll check this PR any way. It would have taken less time if @jack-theripper would remove non relevant changes, such as "private → protected", "Accept" header and others.

**Pull request closed without merging**